presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,440 @@
+                return new JoinEnumerationResult(Optional.empty(), Optional.empty());
+            }
+            List<Expression> joinFilters = joinPredicates.stream()
+                    .filter(predicate -> !isJoinEqualityCondition(predicate))
+                    .collect(toImmutableList());
+
+            Set<Symbol> requiredJoinSymbols = ImmutableSet.<Symbol>builder()
+                    .addAll(outputSymbols)
+                    .addAll(DependencyExtractor.extractUnique(joinPredicates))
+                    .build();
+
+            JoinEnumerationResult leftResult = getJoinSource(
+                    idAllocator,
+                    ImmutableList.copyOf(leftSources),
+                    requiredJoinSymbols.stream().filter(leftSymbols::contains).collect(toImmutableList()));
+            if (leftResult.cost.isPresent() && leftResult.cost.get().isUnknown()) {
+                return new JoinEnumerationResult(Optional.of(UNKNOWN_COST), Optional.empty());
+            }
+            PlanNode left = leftResult.planNode.orElseThrow(() -> new IllegalStateException("no planNode present"));
+            JoinEnumerationResult rightResult = getJoinSource(
+                    idAllocator,
+                    ImmutableList.copyOf(rightSources),
+                    requiredJoinSymbols.stream()
+                            .filter(symbol -> !leftSymbols.contains(symbol))
+                            .collect(toImmutableList()));
+            PlanNode right = rightResult.planNode.orElseThrow(() -> new IllegalStateException("no planNode present"));
+            if (rightResult.cost.isPresent() && rightResult.cost.get().isUnknown()) {
+                return new JoinEnumerationResult(Optional.of(UNKNOWN_COST), Optional.empty());
+            }
+
+            // sort output symbols so that the left input symbols are first
+            List<Symbol> sortedOutputSymbols = Stream.concat(left.getOutputSymbols().stream(), right.getOutputSymbols().stream())
+                    .filter(outputSymbols::contains)
+                    .collect(toImmutableList());
+
+            return setJoinNodeProperties(new JoinNode(
+                    idAllocator.getNextId(),
+                    INNER,
+                    left,
+                    right,
+                    joinConditions,
+                    sortedOutputSymbols,
+                    joinFilters.isEmpty() ? Optional.empty() : Optional.of(and(joinFilters)),
+                    Optional.empty(),
+                    Optional.empty(),
+                    Optional.empty()));
+        }
+
+        private JoinEnumerationResult getJoinSource(PlanNodeIdAllocator idAllocator, List<PlanNode> nodes, List<Symbol> outputSymbols)
+        {
+            PlanNode planNode;
+            if (nodes.size() == 1) {
+                planNode = getOnlyElement(nodes);
+                Expression filter = combineConjuncts(allInference.generateEqualitiesPartitionedBy(outputSymbols::contains).getScopeEqualities());
```

Comment:
I wonder it we shouldn't have any check that all initial filters were somehow consumed, but that is probably difficult since the final filters are different because of inference. I guess we can skip it.